controllers/clustercache/cluster_cache.go

```diff
@@ -0,0 +1,650 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package clustercache
+
+import (
+	"context"
+	"crypto/rsa"
+	"fmt"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/rest"
+	"k8s.io/utils/set"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	"sigs.k8s.io/cluster-api/util/predicates"
+)
+
+// Options defines the options to configure a ClusterCache.
+type Options struct {
+	// SecretClient is the client used to access secrets of type secret.Kubeconfig, i.e. Secrets
+	// with the following name format: "<cluster-name>-kubeconfig".
+	// Ideally this is a client that caches only kubeconfig secrets, it is highly recommended to avoid caching all secrets.
+	// An example on how to create an ideal secret caching client can be found in the core Cluster API controller main.go file.
+	SecretClient client.Reader
+
+	// WatchFilterValue is the label value used to filter events prior to reconciliation.
+	WatchFilterValue string
```

Comment:
Up until now we didn't have that requirement for workload clusters. I would not add it for now.

We can make this possible later if necessary. But than I would prefer directly exposing the corresponding cache.Option from CR instead of the magic we have in main.go

https://github.com/kubernetes-sigs/cluster-api/blob/84a33475f72909b083b5f069b64713b5a4d45e0e/main.go#L322

(i.e. I would expose DefaultNamespaces with the same type as in CR cache.Options instead of watchNamespaces)

Reply:
Regarding the cache that is used by the ClusterCache reconciler itself we're good, because it just uses the one we configure in main.go for manager